crates/engine/tree/src/tree/root.rs

```diff
@@ -118,31 +218,389 @@ where
 
     /// Handles state updates.
     fn on_state_update(
-        _view: &reth_provider::providers::ConsistentDbView<impl Send + 'static>,
-        _input: &std::sync::Arc<reth_trie::TrieInput>,
-        _state: EvmState,
+        view: ConsistentDbView<Factory>,
+        input: Arc<TrieInput>,
+        update: EvmState,
+        state: &mut HashedPostState,
+        proof_sequence_number: u64,
+        state_root_message_sender: Sender<StateRootMessage>,
     ) {
-        // Default implementation of state update handling
-        // TODO: calculate hashed state update and dispatch proof gathering for it.
+        let mut hashed_state_update = HashedPostState::default();
+        for (address, account) in update {
+            if account.is_touched() {
+                let hashed_address = keccak256(address);
+
+                let destroyed = account.is_selfdestructed();
+                hashed_state_update.accounts.insert(
+                    hashed_address,
+                    if destroyed || account.is_empty() { None } else { Some(account.info.into()) },
```

Comment:
account can be destroyed and recreated, worth checking again if this line is correct